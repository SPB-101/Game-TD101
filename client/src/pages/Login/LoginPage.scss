.login-page {
  &__title {
    text-align: center;
    font-family: "Hacked", Arial, Helvetica, sans-serif;
    letter-spacing: 0.1em;
    font-size: var(--text-size-xxl);
    text-shadow: var(--shadow-glitch);
  }

  &__form {
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin-bottom: 10px;
    margin-top: 30px;
    gap: 16px;
    padding: 44px 16px 16px 16px;
    border: var(--border) transparent;

    &_error {
      border-color: var(--color-red);
    }
  }

  &__link {
    text-align: center;
  }

  &__error-text {
    position: absolute;
    top: 0;
    max-width: calc(100% - 30px);
    transform: translate(0, -50%);
    text-align: center;
    font-size: var(--text-size-xs);
    background-color: var(--color-gray-1);
    padding: 4px;
    border: var(--border) var(--color-red);

    span {
      position: relative;
      animation-name: shake;
      animation-duration: 100ms;
      transform-origin: 50% 50%;
      animation-iteration-count: 2;
      animation-timing-function: linear;
    }
  }
}

@keyframes shake {
  10%,
  90% {
    transform: translate(-1px);
  }
  20%,
  80% {
    transform: translate(1px);
  }
  30%,
  50%,
  70% {
    transform: translate(-1px);
  }
  40%,
  60% {
    transform: translate(1px);
  }
}
