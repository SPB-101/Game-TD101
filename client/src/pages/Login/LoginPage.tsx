import React, { useCallback } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { Form, Field } from "react-final-form";
import classNames from "classnames";

import { Wrapper } from "../../component/Wrapper";
import { TextField } from "../../component/TextField";
import { Button } from "../../component/Button";
import { Loader } from "../../component/Loader";

import {
  getErrorMessage,
  getIsLoading,
} from "../../store/selectors/widgets/loginPage";
import { State } from "../../store/reducers";
import { fetchLogin } from "../../store/thunks/widgets/login";

import { range } from "../../utils/validation/rules";
import { validate } from "../../utils/validate";

import "./LoginPage.scss";

import type { Props } from "./types";

const ruelesFields = [
  { field: "login", callback: (v: string | number) => range(v, 3) },
  { field: "password" },
];

export const LoginPageBlock = ({
  errorMessage,
  fetchLoginThunk,
  isLoading,
}: Props): JSX.Element => {
  const { t } = useTranslation();

  const onSubmit = useCallback((values: Record<string, string>) => {
    return fetchLoginThunk({
      login: values.login,
      password: values.password,
    });
  }, []);

  return (
    <Wrapper className="login-page" size="m">
      <h1 className="login-page__title">{t("nameGame")}</h1>
      <Form
        onSubmit={onSubmit}
        validate={validate(ruelesFields)}
        render={({ handleSubmit }) => (
          <form
            className={classNames("login-page__form", {
              ["login-page__form_error"]: errorMessage,
            })}
            onSubmit={handleSubmit}
          >
            {isLoading && <Loader />}
            {errorMessage && (
              <div className="login-page__error-text">
                <span>{errorMessage}</span>
              </div>
            )}

            <Field name="login">
              {({ input, meta }) => (
                <TextField
                  {...input}
                  error={meta.error && meta.touched ? meta.error : ""}
                  name="login"
                  label={t("login")}
                  placeholder="login"
                />
              )}
            </Field>

            <Field name="password">
              {({ input, meta }) => (
                <TextField
                  {...input}
                  error={meta.error && meta.touched ? meta.error : ""}
                  name="password"
                  type="password"
                  label={t("password")}
                  placeholder=""
                />
              )}
            </Field>

            <Button type="submit" disabled={isLoading}>
              {t("login")}
            </Button>

            <Button disabled={true}>{t("loginYandex")}</Button>
          </form>
        )}
      />
      <Link className="login-page__link" to="/registration">
        {t("registration")}
      </Link>
    </Wrapper>
  );
};

const mapStateToProps = (state: State) => ({
  errorMessage: getErrorMessage(state),
  isLoading: getIsLoading(state),
});

const mapDispatchToProps = { fetchLoginThunk: fetchLogin };

export const LoginPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginPageBlock);
