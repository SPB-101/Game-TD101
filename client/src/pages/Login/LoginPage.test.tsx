import React from "react";
import renderer from "react-test-renderer";
import { LoginPageBlock } from "./LoginPage";
import { StaticRouter } from "react-router";

describe("<LoginPageBlock />", () => {
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <StaticRouter location="/login">
          <LoginPageBlock
            isLoading={false}
            errorMessage=""
            fetchLoginThunk={() => {
              /**/
            }}
          />
        </StaticRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
