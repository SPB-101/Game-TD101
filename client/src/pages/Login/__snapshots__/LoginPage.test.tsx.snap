// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPageBlock /> should render correctly 1`] = `
<div
  className="wrapper wrapper_size--m login-page"
>
  <h1
    className="login-page__title"
  >
    nameGame
  </h1>
  <form
    className="login-page__form"
    onSubmit={[Function]}
  >
    
    <label
      className="text-field"
      htmlFor="login"
    >
      <input
        className="text-field__input"
        name="login"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="login"
        value=""
      />
      <div
        className="text-field__label"
      >
        login
      </div>
      <span
        className="text-field__error"
      >
        
      </span>
    </label>
    <label
      className="text-field text-field_password"
      htmlFor="password"
    >
      <input
        className="text-field__input"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        type="password"
        value=""
      />
      <div
        className="text-field__label"
      >
        password
      </div>
      <span
        className="text-field__error"
      >
        
      </span>
    </label>
    <button
      className="button"
      onClick={[Function]}
    >
      login
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      loginYandex
    </button>
  </form>
  <a
    className="login-page__link"
    href="/registration"
    onClick={[Function]}
  >
    registration
  </a>
</div>
`;
